### PR TITLE
Only generate bodies for properties if they are readonly, writeonly o…

### LINKED ISSUE #5898: Visual Basic Auto-implement interface behaviour changed
If you have code like this:

``` VB
Public Class BandEntry
    Implements IBandEntry
End Class

Friend Interface IBandEntry
    Property BandStart As Decimal
    Property BandTypeID As Integer
    Property Rate As Decimal
End Interface
```

In VS2013 when I used the auto-implement interface feature (in VB, press return after implements statement, body of class would be auto populated) it would produce out like the following: 

``` VB
Public Class BandEntry
    Implements IBandEntry

    Public Property BandStart As Decimal Implements IBandEntry.BandStart

    Public Property BandTypeID As Integer Implements IBandEntry.BandTypeID

    Public Property Rate As Decimal Implements IBandEntry.Rate
End Class
```

In Visual Studio 2015 it produces output like this: 

``` VB
Public Class BandEntry
    Implements IBandEntry

    Public Property BandStart As Decimal Implements IBandEntry.BandStart
        Get
            Throw New NotImplementedException()
        End Get
        Set(value As Decimal)
            Throw New NotImplementedException()
        End Set
    End Property

    Public Property BandTypeID As Integer Implements IBandEntry.BandTypeID
        Get
            Throw New NotImplementedException()
        End Get
        Set(value As Integer)
            Throw New NotImplementedException()
        End Set
    End Property

    Public Property Rate As Decimal Implements IBandEntry.Rate
        Get
            Throw New NotImplementedException()
        End Get
        Set(value As Decimal)
            Throw New NotImplementedException()
        End Set
    End Property
End Clas
```

Is it possible to have the old behaviour back either by default or as an option? We commonly use interfaces to specify data types between our presentation and view layers, and a lot of the time all we need us the auto-property implementation. Since this change I need to go through the generated code removing 90% of it every time. I appreciate others might have to do the same in reverse, but I'd suggest they'd need to go through the generated code replacing the bodies anyway, so there's less wasted effort that way. Making it an option, or perhaps having two choices from a lightbulb menu would make both camps happy?

### PR DESCRIPTION
**Pull request by jmarolf**
…r an indexer

Fixes #5898

### CODE REVIEW

**Comment by jmarolf** [comment]
Tagging @dotnet/mlangideTeam

File: src/Workspaces/Core/Portable/CodeGeneration/AbstractCodeGenerationService.cs [hunk 1]
```diff
@@ -233,7 +233,7 @@ protected static T Cast<T>(object value)
                     currentDestination = member.TypeSwitch(
                         (IEventSymbol @event) => this.AddEvent(currentDestination, @event, options, availableIndices),
                         (IFieldSymbol field) => this.AddField(currentDestination, field, options, availableIndices),
-                        (IPropertySymbol property) => this.AddProperty(currentDestination, property, options, availableIndices),
+                        (IPropertySymbol property) => this.AddProperty(currentDestination, property, ShouldGenerateMethodBody(options, property), availableIndices),
```
Comment (brettfo) [inline src/Workspaces/Core/Portable/CodeGeneration/AbstractCodeGenerationService.cs]: Why not pass the same parameters as before and instead call `ShouldGenerateMethodMody()` inside of `AddProperty`?

File: src/Workspaces/Core/Portable/CodeGeneration/AbstractCodeGenerationService.cs [hunk 2]
```diff
@@ -276,6 +276,9 @@ protected static T Cast<T>(object value)
             return currentDestination;
         }
 
+        private static CodeGenerationOptions ShouldGenerateMethodBody(CodeGenerationOptions options, IPropertySymbol property) =>
+            options.GenerateMethodBodies ? options.WithGenerateMethodBodies(property.IsIndexer || property.IsWriteOnly || property.IsReadOnly || property.Parameters.Length > 0) : options;
```
Comment (Pilchie) [inline src/Workspaces/Core/Portable/CodeGeneration/AbstractCodeGenerationService.cs]: Is this right in C#?  You want to generate a body instead of a new getter only autoproperty if `IsReadOnly`?

Comment (jmarolf) [inline src/Workspaces/Core/Portable/CodeGeneration/AbstractCodeGenerationService.cs]: I am making a conservative choice in assuming that if the user has a readonly or writeonly property that they may want to add some custom logic, so we generate a body.  We obviously could generate a getter-only autoprop, but the user would need to write additional code to initialize it anyway.  Do you disagree? would users want getter-only autoproperties to be generated?

Comment (Pilchie) [inline src/Workspaces/Core/Portable/CodeGeneration/AbstractCodeGenerationService.cs]: Ideally, I think we'd want to have an option that allows the user to control whether or not they want the IDE to generate autoproperties when possible, and a refactoring to make "convert to autprop" to expand an autoprop into a full property.  Tagging @DustinCampbell and @CyrusNajmabadi explicitly to get their thoughts.

File: src/EditorFeatures/CSharpTest/Diagnostics/ImplementInterface/ImplementInterfaceTests.cs [hunk 3]
```diff
@@ -192,7 +192,7 @@ public void TestConflictingProperties()
         {
             Test(
 @"class Test : [|I1|] { int Prop { get ; set ; } } interface I1 { int Prop { get ; set ; } } ",
-@"using System ; class Test : I1 { int I1 . Prop { get { throw new NotImplementedException ( ) ; } set { throw new NotImplementedException ( ) ; } } int Prop { get ; set ; } } interface I1 { int Prop { get ; set ; } } ");
+@"class Test : I1 { int I1 . Prop { get ; set ; } int Prop { get ; set ; } } interface I1 { int Prop { get ; set ; } } ");
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/CSharpTest/Diagnostics/ImplementInterface/ImplementInterfaceTests.cs]: This is qite a change.  Is @DustinCampbell ok with this?  It was a big part of how we generated code before that all the code would throw by default when you used it.  Now we're not doing that.

File: src/Workspaces/Core/Portable/CodeGeneration/AbstractCodeGenerationService.cs (hunk 1 above)
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/CodeGeneration/AbstractCodeGenerationService.cs]: agreed.

File: src/Workspaces/Core/Portable/CodeGeneration/AbstractCodeGenerationService.cs (hunk 2 above)
Comment (CyrusNajmabadi) [inline src/Workspaces/Core/Portable/CodeGeneration/AbstractCodeGenerationService.cs]: Agreed.  I think this should be a user option.

File: src/EditorFeatures/CSharpTest/Diagnostics/ImplementInterface/ImplementInterfaceTests.cs (hunk 3 above)
Comment (DustinCampbell) [inline src/EditorFeatures/CSharpTest/Diagnostics/ImplementInterface/ImplementInterfaceTests.cs]: I'm not sure -- though we should discuss this sort of design change before making them. I'm not a huge fan of dramatically changing how features work in an update.

@jmarolf, could you add context from the original bug so we know why this change is necessary? Better yet, could you port the relevant portion of the bug to GitHub?

File: src/Workspaces/Core/Portable/CodeGeneration/AbstractCodeGenerationService.cs (hunk 2 above)
Comment (DustinCampbell) [inline src/Workspaces/Core/Portable/CodeGeneration/AbstractCodeGenerationService.cs]: Absolutely. Agreed 100%.

File: src/EditorFeatures/CSharpTest/Diagnostics/ImplementInterface/ImplementInterfaceTests.cs (hunk 3 above)
Comment (jmarolf) [inline src/EditorFeatures/CSharpTest/Diagnostics/ImplementInterface/ImplementInterfaceTests.cs]: @DustinCampbell ported the bug to github https://github.com/dotnet/roslyn/issues/5898.  The problem is that VB used to prefer to generate autoprops and we broke those people,  The minimal fix here would be to only generate autoprops for VB users and let C# folks delete all those property bodies.

Comment (CyrusNajmabadi) [inline src/EditorFeatures/CSharpTest/Diagnostics/ImplementInterface/ImplementInterfaceTests.cs]: I'd much rather the languages behave consistency, with options to let users opt into differing behavior (*if* we think there's enough value in different defaults).  For auto-props I think there is *enough* value.  It feels somewhat archaic for us to so often to be using older forms in the language when we could be moving to more modern and (IMO, more pleasant) constructs.

Comment (jmarolf) [inline src/EditorFeatures/CSharpTest/Diagnostics/ImplementInterface/ImplementInterfaceTests.cs]: @CyrusNajmabadi I agree.  So we should have an option for this. My initial assumption was that we should generate autoprops  when possible unless they are readonly or writeonly.  Do we think that is a good option and have the other option be to generate property bodies?

Comment (DustinCampbell) [inline src/EditorFeatures/CSharpTest/Diagnostics/ImplementInterface/ImplementInterfaceTests.cs]: Yes, we should add an option. @CyrusNajmabadi and I were just talking a few days ago about needing a "Prefer auto-properties" option to help control the "Use auto-property" diagnostic/fix.

Comment (jmarolf) [inline src/EditorFeatures/CSharpTest/Diagnostics/ImplementInterface/ImplementInterfaceTests.cs]: Alright, standby for options.

**Comment by DustinCampbell** [comment]
What's the status of this PR? Are you waiting on anything else from us?
